### ISSUE TITLE
Incomplete initialization of user-$UID.slice disturbs systemd-run --user -t -p MemoryHigh=1G

### ISSUE THREAD

**Issue by ghost** [systemd/systemd#9512]
**systemd version the issue has been seen with**

v239

**Used distribution**

Debian sid
Filed as downstream bug report at [https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=903011](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=903011)
I reported the above issue to the Debian Bug Tracking System and Michael Biebl requested me to report it here.

**Expected behaviour you didn't see**

I can impose CGroup V2 memory limitation by `systemd-run --user -t -p "MemoryHigh=1G" /bin/bash`

Within the bash, I see

```
$ cat /sys/fs/cgroup/user.slice/user-1005.slice/user@1005.service/run-u0.service/memory.high 
1073741824
```

and the memory limitation seems working. It has been convenient for preventing memory-eating applications from eating all the physical memory.

I verified the above behavior in Ubuntu 18.04 LTS running in systemd-nspawn. Its systemd package version is 237-3ubuntu10. Host Linux is also Ubuntu 18.04 LTS.

**Unexpected behaviour you saw**

When I do the same in Debian Unstable (in a container), `memory.high` does not appear and the memory limitation does not seem working. Systemd version in Debian sid is 239.

**Steps to reproduce the problem**

The configuration of Debian Unstable is almost the same as Ubuntu 18.04 running in a container. Specifically,

* `dbus-session-user` package is chosen to enable `systemd-run --user`.

* `/etc/systemd/system/user@.service.d/delegate.conf` has the following
```
[Service]
Delegate=memory pids cpu io
```

* Host Linux has the following kernel command line to let systemd use CGroup V2. `systemd.unified_cgroup_hierarchy=1`

* Linux kernel package version is 4.15.0-24-generic

**Comment by ryutaroh-matsumoto** [comment]
I found a way to avoid this bug, and let systemd 239 operate as expected, as follows:

1. Prepare the following two config files

```
# as /etc/systemd/system/user-.slice.d/my-default.conf
[Slice]
Delegate=memory pids cpu io
MemoryHigh=32G
```

```
# as /etc/systemd/system/user@.service.d/delegate.conf
Delegate=memory pids cpu io
MemoryHigh=32G
```

2. Reboot the system or the container (I tested this in a systemd-nspawn container).
3. `touch /etc/systemd/system/user-.slice.d/my-default.conf /etc/systemd/system/user@.service.d/delegate.conf` as root
4. `systemctl daemon-reload` as root. 

By the above steps, the problem described in this issue disappears, like
```
ryutaroh@gtx1080ti-unstable:~$ systemd-run --user --scope -p MemoryHigh=3G /bin/bash
Running scope as unit: run-r1d00f909a93c4635a4d6dce770d0bbde.scope
ryutaroh@gtx1080ti-unstable:~$ cat /sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/run-r1d00f909a93c4635a4d6dce770d0bbde.scope/memory.high
3221225472
```

In addition, this workaroud also helps the issue [9502](https://github.com/systemd/systemd/issues/9502), as
```
ryutaroh@gtx1080ti-unstable:~$ systemd-run --user --no-block -p "MemoryAccounting=yes" /bin/bash -c "sleep 1000"
Running as unit: run-r302f6c26d9f94b75972267de3c3cb2b8.service
ryutaroh@gtx1080ti-unstable:~$ systemctl --user --no-pager -l status run-r302f6c26d9f94b75972267de3c3cb2b8.service
● run-r302f6c26d9f94b75972267de3c3cb2b8.service - /bin/bash -c sleep 1000
   Loaded: loaded (/run/user/1000/systemd/transient/run-r302f6c26d9f94b75972267de3c3cb2b8.service; transient)
Transient: yes
   Active: active (running) since Wed 2018-07-11 18:16:30 JST; 17s ago
 Main PID: 373 (sleep)
    Tasks: 1 (limit: 4915)
   Memory: 332.0K
      CPU: 5ms
   CGroup: /user.slice/user-1000.slice/user@1000.service/run-r302f6c26d9f94b75972267de3c3cb2b8.service
           └─373 sleep 1000

Jul 11 18:16:30 gtx1080ti-unstable.fujii.nuee.nagoya-u.ac.jp systemd[339]: Started /bin/bash -c sleep 1000.
```

I deeply wonder why this works....

**Comment by ryutaroh-matsumoto** [comment]
My last comment was checked with systemd 239 running in systemd-nspawn, but I found that the above method was insufficient to avoid this issue in systemd 239 running directly on Linux without a container. I checked the below steps with systemd 239 running on Debian unstable and Linux 4.17 (pulled from Debian experimental).

1. Prepare the following four configuration files:

```
# as /etc/systemd/system/-.slice
# and /etc/systemd/system/user.slice.d/delegate-1T.conf
[Slice]
Delegate=memory pids cpu io
MemoryHigh=1T
```

```
# as /etc/systemd/system/user-.slice.d/delegate-slice.conf
[Slice]
Delegate=memory pids cpu io
MemoryHigh=48G
```

```
# as /etc/systemd/system/ user@.service.d/ delegate-service.conf
[Service]
Delegate=memory pids cpu io
MemoryHigh=48G
```

2.  Run `cd /etc/systemd/system; touch -- -.slice user.slice.d/delegate-1T.conf user-.slice.d/delegate-slice.conf user@.service.d/delegate-service.conf` as root. This seems necessary to let systemd to reload those files at the next step.

3. Run `systemctl daemon-reload`

4. From the next login, this issue and [9502](https://github.com/systemd/systemd/issues/9502) should not appear with systemd 239.

**Comment by ryutaroh-matsumoto** [comment]
I found that the workaround in my last comment was unreliable, i.e., it sometimes works and sometimes fails. The root cause of this issue and [9502](https://github.com/systemd/systemd/issues/9502) seems that `user.slice` does not have CGroup V2 controllers. I found that the following workaround always _fix_ this issue and [9502](https://github.com/systemd/systemd/issues/9502). But it does not seem a right approach.

1. Make the following simple shell script as `/root/delegate.sh`

```
#!/bin/bash
echo "+memory +pids +cpu +io" >>/sys/fs/cgroup/cgroup.subtree_control
echo "+memory +pids +cpu +io" >>/sys/fs/cgroup/user.slice/cgroup.subtree_control
```

2. Install a service file so that the above script is always executed at the boot time, for example, 

```
[Unit]
Description=Fix cgroup controllers
After=user.slice
Requires=user.slice

[Service]
Type=oneshot
ExecStart=/bin/bash /root/delegate.sh
KillMode=process

[Install]
WantedBy=multi-user.target
```

**Comment by ryutaroh-matsumoto** [comment]
In case of this issue being distro-specific, I created a disk image for qemu/kvm [here](https://drive.google.com/file/d/1rsIu6NFwU9W6JMgA-vKPBi-nNcSQmNqM/view) so that users of other distros can reproduce and investigate this issue. The file is tar+gzipped QCOW2 image file of 600 MB, and it can be used with `Import existing disk image` in `Create a new virtual machine` window of `virt-manager`. In the disk image,

* Architecture is x86-64 and systemd package version is 239-5.
* Debian buster (testing) as of 15 July 2018 with `standard` and `openssh server` are selected in the installation process of [Debian buster alpha 3 installer](https://cdimage.debian.org/cdimage/buster_di_alpha3/amd64/iso-cd/).
* User `root` has password `root` and user `guest2` has password `guest2`.
* Configuration files are untouched except `/etc/default/grub` to enable the unified cgroup hirerarchy.
* Keyboard config is US and it can be changed by `dpkg-reconfigure keyboard-configuration`.
* Ssh login is accepted but root ssh login is not allowed.

If you find this issue being a problem of Debian, please let me know.

**Comment by ryutaroh-matsumoto** [comment]
The cuase of this issue seems incomplete initialization of `user-$UID.slice`. More explicit initialization of `user-$UID.slice` _solves_ the issue. For example, prepare the following file as `/etc/systemd/systemd/user-1000.slice` and run `systemctl enable user-1000.slice`. Then UID 1000 does not experience this issue from the next reboot.

```
[Unit]
Description=User Slice of UID 1000
After=systemd-user-sessions.service

[Slice]
TasksMax=33%

[Install]
WantedBy=multi-user.target
```

**Comment by ryutaroh-matsumoto** [comment]
Additional information: This issue disappears if `user@$UID.service` is started by `multi-user.target`. This issue continues to appear after `loginctl enable-linger $UID` is executed by root.

**Comment by poettering** [comment]
Fix waiting in #10894.
